File: problems/39. Combination Sum/main.go
func combinationSum(candidates []int, target int) [][]int {
	var result [][]int

	var helper func(remaining int, start int, combination []int)
	helper = func(remaining int, start int, combination []int) {
		if remaining == 0 {
			uniqueCombination := append([]int{}, combination...)
			result = append(result, uniqueCombination)
			return
		} else if remaining < 0 {
			return
		}

		for i := start; i < len(candidates); i++ {
			combination = append(combination, candidates[i])
			helper(remaining-candidates[i], i, combination)
			combination = combination[:len(combination)-1]
		}
	}

	helper(target, 0, []int{})
	return result
}

File: problems/39. Combination Sum/main.py
class Solution(object):
    def combinationSum(self, candidates, target):
        result = []

        def helper(remaining, start, combinations):
            if remaining == 0:
                result.append(list(combinations))
                return
            elif remaining < 0:
                return

            for i in range(start, len(candidates)):
                combinations.append(candidates[i])
                helper(remaining - candidates[i], i, combinations)
                combinations.pop()

        helper(target, 0, [])
        return result
